src/legacy/reduce.rs

```diff
@@ -0,0 +1,488 @@
+        let v = curr_byte;
+        let mut len = lsb(v as u64, v_len_bits) as usize;
+        if len == (1 << v_len_bits) - 1 {
+            // Read an extra length byte.
+            curr_byte = read_next_byte(&mut is, curr_byte, &mut fsets)?;
+            len += curr_byte as usize;
+        }
+        len += 3;
+
+        // Read the W byte, which together with V gives the distance.
+        curr_byte = read_next_byte(&mut is, curr_byte, &mut fsets)?;
+        let dist = ((v as usize) >> v_len_bits) * 256 + curr_byte as usize + 1;
+
+        debug_assert!(len <= max_len(comp_factor));
+        debug_assert!(dist as usize <= max_dist(comp_factor));
+
+        // Output the back reference.
+        if len <= uncomp_len - dst.len() && dist as usize <= dst.len() {
+            // Enough room and no implicit zeros; chunked copy.
+            lz77_output_backref(dst, dist as usize, len);
+        } else {
+            // Copy, handling overlap and implicit zeros.
+            for _i in 0..len {
+                if dist as usize > dst.len() {
+                    dst.push_back(0);
+                    continue;
+                }
+                dst.push_back(dst[dst.len() - dist as usize]);
+            }
+        }
+    }
+
+    *src_used = is.bytes_read();
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct ReduceDecoder<R> {
+    compressed_reader: R,
+    uncompressed_size: u64,
+    stream_read: bool,
+    comp_factor: u8,
+    stream: VecDeque<u8>,
+}
+
+impl<R: Read> ReduceDecoder<R> {
+    pub fn new(inner: R, uncompressed_size: u64, comp_factor: u8) -> Self {
+        ReduceDecoder {
+            compressed_reader: inner,
+            uncompressed_size,
+            stream_read: false,
+            comp_factor,
+            stream: VecDeque::new(),
+        }
+    }
+
+    pub fn finish(mut self) -> Result<VecDeque<u8>> {
+        copy(&mut self.compressed_reader, &mut self.stream)?;
+        Ok(self.stream)
+    }
+}
+
+impl<R: Read> Read for ReduceDecoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if !self.stream_read {
+            self.stream_read = true;
+            let mut compressed_bytes = Vec::new();
+            if let Err(err) = self.compressed_reader.read_to_end(&mut compressed_bytes) {
+                return Err(err.into());
+            }
+            let mut src_used = 0;
+            hwexpand(
+                &compressed_bytes,
+                compressed_bytes.len(),
+                self.uncompressed_size as usize,
+                self.comp_factor,
+                &mut src_used,
+                &mut self.stream,
+            )?;
+        }
+        let bytes_read = self.stream.len().min(buf.len());
+        buf[..bytes_read].copy_from_slice(&self.stream.drain(..bytes_read).collect::<Vec<u8>>());
+        Ok(bytes_read)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::VecDeque;
+
+    use crate::legacy::reduce::{follower_idx_bw, max_dist};
+
+    use super::hwexpand;
+
```

Comment:
Move this to a file in tests/data and use `include_bytes!`.

Reply:
k